apps/aechannel/src/aesc_utils.erl

```diff
@@ -791,3 +851,46 @@ add_call(Call0, TxHash, Trees, Env) ->
                        {ok, aec_trees:trees(), aetx_env:env()}.
 add_event(Trees, ChannelPubKey, Env) ->
     {ok, Trees, aetx_env:tx_event({channel, ChannelPubKey}, Env)}.
+
+count_authentications(SignedTx) ->
+    count_authentications(SignedTx, 0).
+
+count_authentications(SignedTx, Cnt) ->
+    case aetx:specialize_callback(aetx_sign:tx(SignedTx)) of
+        {aega_meta_tx, InnerSignedTx} -> 
+            count_authentications(aega_meta_tx:tx(InnerSignedTx), Cnt + 1);
+        {_, _} -> % most inner tx
+            Cnt + length(aetx_sign:signatures(SignedTx))
+    end.
+
+-spec channel_pubkey(aetx_sign:signed_tx()) -> {ok, aec_keys:pubkey()} |
+                                               {error, not_channel_tx}.
+channel_pubkey(SignedTx) ->
+    case aetx:specialize_callback(aetx_sign:innermost_tx(SignedTx)) of
+        %% initial channel id depends on the auth
+        {aesc_create_tx, Txi} ->
+            Initiator = aesc_create_tx:initiator_pubkey(Txi),
+            Responder = aesc_create_tx:responder_pubkey(Txi),
+            InitiatorAuthId = get_nonce_or_auth_id(SignedTx,
+                                                   Initiator),
+            PK = aesc_channels:pubkey(Initiator,
+                                      InitiatorAuthId,
+                                      Responder),
+            {ok, PK};
+        %% Likely only channel txs have a channel_id/1 callback, so prepare for
+        %% 'undef' exceptions.
+        {Mod, Txi} ->
+            try {ok, Mod:channel_pubkey(Txi)}
+            catch error:_ -> {error, not_channel_tx}
+            end
+    end.
+
+get_nonce_or_auth_id(SignedTx, Signer) ->
+    case aetx:specialize_callback(aetx_sign:tx(SignedTx)) of
+        {aega_meta_tx, MetaTx} ->
+            case aega_meta_tx:ga_pubkey(MetaTx) =:= Signer of
+                true -> aega_meta_tx:auth_id(MetaTx);
```

Comment:
This seems to take the auth id of **outermost** ga meta tx related to the account, while an on-chain processing of tx takes the **innermost** one (see [`aega_meta_tx:process`](https://github.com/aeternity/aeternity/blob/996df3264e07c6b6f5b3f9b8f5c3d10ca2dd1ffe/apps/aega/src/aega_meta_tx.erl#L187)). Is this intended?

I am still getting my head around this, and this may have no impact if multiple ga meta tx for same account prevented somewhere else. Though there may be room for opportunity in readability (maybe in other point in code) hence this note.

Reply:
Different transaction types have different `origin` and the `origin` of `channel_create_transaction` is the `initiator`. This function is intended to get the `initiator`'s nonce or auth_id out from `channel_create_transaction` so it can later on be used in `channel_id` computation. In case that the outermost transaction indeed is a meta one, it could be either the `initiator` or the `responder` authentication method, thus the `aega_meta_tx:ga_pubkey(MetaTx) =:= Signer` check - we return only the `Signer`'s `auth_id`.

What botters me now is the other option - if none of the participants is a GA, the funciton returns `Mod:nonce(Tx)`. This could return unexpected result if the `get_nonce_or_auth_id/2` function is called for the `responder` instead.

I am to rework and rename it to express more clearly the intention regarding the transaction type and the participant.

Reply:
Addressed

Reply:
Change is improvement thanks.

Doubt re outermost-vs-innermost still stays. For the same account-meant-to-be-authenticated, multiple ga meta txs can be nested within each other.